### PR TITLE
Niad-2455: JMeter performance load test

### PR DESCRIPTION
**Pull request by ole4ryb**
## What

JMeter load tests were run to confirm PS Adaptor performance ability to handle 2000 transfers load in a short period of time.

## Why

JMeter load tests were run in order to be sure about PS Adaptor performance and stability under extreme load.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Internal change (non-breaking change with no effect on the functionality affecting end users)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [] I have updated the [Changelog](/CHANGELOG.md) with details of my change in the UNRELEASED section if this change will affect end users
- [ ] A corresponding change has been made to the [Mapping Documentation repository][mapping-docs]

[mapping-docs]: https://github.com/NHSDigital/patient-switching-adaptors-mapping-documentation

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -137,3 +137,28 @@ In particular this means that this project may not depend on GPL-licensed or AGP
...
+
+## Performance
+The performance of PS Adaptor was tested with JMeter tool.
+There was a series of tests run with the following setup and parameters:
+- PS Adaptor is run in ECS AWS environment with 4 CPUs and 16 GB memory.
+- For the message queue mq.m5.xlarge host type was used
+- RDS DB host type was set to db.t3.xlarge
+
+The test used 2000 transfers which were split into 5 batches of 400 transfers.
+Pause time between transfers was set to 1 sec and the socket timeout was to 2 minutes.
```
Comment (adrianclay) [inline README.md:149]: I'm still unsure as to whether 120seconds is realistic. Perhaps we should look at the Splunk logs to see what timeout Spine uses and align to that.

File: test-suite/non-functional-tests/test-scenario-6/config-template.properties [hunk 2]
```diff
@@ -1,22 +1,21 @@
...
 
 # * Connection Properties *
-facadePort = {{facade-port}}
-facadeUrl = {{facade-url}}
-
-inboundUrl = {{inbound-url}}
-inboundPort = {{inbound-port}}
-
+facadePort = 8081
+facadeUrl = nia-gp2gp-elb-ebd5089784c3d1ec.elb.eu-west-2.amazonaws.com
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-6/config-template.properties:5]: I think it'd be sensible to tear down the infrastructure once we're Done with this ticket, so I wouldn't commit the temporary URLs.

File: test-suite/non-functional-tests/test-scenario-6/README.md [hunk 3]
```diff
@@ -25,6 +25,42 @@ To start the tests, run the following:
 ./start-docker.test.sh
 ```
 
+
+### Load Testing
```
Comment (adrianclay) [inline test-suite/non-functional-tests/test-scenario-6/README.md:29]: So this folder has 6 different "test-scenario" folders. What were your thoughts on that layout @ole4ryb / @azam145 ?

**Comment by adrianclay** [comment]
Is there any PRs for the [NHSDigital/patient-switching-standard-adaptors-architecture](https://github.com/NHSDigital/patient-switching-standard-adaptors-architecture) repo?

File: README.md [hunk 4]
```diff
@@ -137,3 +137,28 @@ In particular this means that this project may not depend on GPL-licensed or AGP
 as these would violate the terms of those libraries' licenses.
 
 The contents of this repository are protected by Crown Copyright (C).
+
+## Performance
```
Comment (adrianclay) [inline README.md:141]: I like this section, but I think we should include a bit more information about what was included in each "trasnfer".

File: README.md [hunk 5]
```diff
@@ -137,3 +137,28 @@ In particular this means that this project may not depend on GPL-licensed or AGP
...
+The performance of PS Adaptor was tested with JMeter tool.
+There was a series of tests run with the following setup and parameters:
+- PS Adaptor is run in ECS AWS environment with 4 CPUs and 16 GB memory.
+- For the message queue mq.m5.xlarge host type was used
+- RDS DB host type was set to db.t3.xlarge
```
Comment (adrianclay) [inline README.md:146]: I think we should mention that the MHS adaptor was also being run, but that Spine responses were being simulated by JMeter.

File: test-suite/non-functional-tests/test-scenario-6/config-template.properties (hunk 2 above)
Comment (ole4ryb) [reply test-suite/non-functional-tests/test-scenario-6/config-template.properties:5]: ok, thanks, yes, we will do that...

File: test-suite/non-functional-tests/test-scenario-6/README.md (hunk 3 above)
Comment (ole4ryb) [reply test-suite/non-functional-tests/test-scenario-6/README.md:29]: So the layout changes are a bit different to the main purpose of this PR which was about testing of PS adapter performance. We will do in a separate PR.

File: README.md [hunk 6]
```diff
@@ -137,3 +137,35 @@ In particular this means that this project may not depend on GPL-licensed or AGP
 as these would violate the terms of those libraries' licenses.
 
 The contents of this repository are protected by Crown Copyright (C).
+
+## Performance
+The performance of PS Adaptor was tested with JMeter tool. 
+The use case tested was the simulation of the patient transfer request. 
+This was tested by sending EHR record requests to the PS Adaptor and we expected to receive a bundle back.
```
[the remark this answers is not in the document]
Comment (ole4ryb) [reply README.md:144]: For the bundle the test only checked the status code. The test didn't measure the number of returned resources.

[the remark this answers is not in the document]
Comment (adrianclay) [reply README.md:144]: Yep, but you can imagine that a load test where transferring patients with 5TB of data is different to a load test where transferring patients with 1KB of data.

As a reader of this section, I'd like to understand the scale of the transfer.

[the remark this answers is not in the document]
Comment (ole4ryb) [reply README.md:144]: Np, I will include the info about the attachment that is used for the patient transfer workflow.

**Review by adrianclay — DISMISSED** [review]

**Review by adrianclay — APPROVED** [review]

**Pull request merged**